sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts

```diff
@@ -111,4 +141,532 @@ export class ConfirmationOfEnrollmentService {
+        const updateResult = await transactionalEntityManager
+          .getRepository(DisbursementSchedule)
+          .createQueryBuilder()
+          .update(DisbursementSchedule)
+          .set({
+            coeStatus: COEStatus.declined,
+            coeUpdatedBy: auditUser,
+            coeUpdatedAt: now,
+            coeDeniedReason: { id: declinedReason.coeDeniedReasonId },
+            coeDeniedOtherDesc: declinedReason.otherReasonDesc,
+            modifier: auditUser,
+            updatedAt: now,
+          })
+          .where("id = :disbursementScheduleId", { disbursementScheduleId })
+          .andWhere("coeStatus = :required", { required: COEStatus.required })
+          .execute();
+
+        if (updateResult.affected !== 1) {
+          throw new Error(
+            `While updating COE status to '${COEStatus.declined}' the number of affected row was bigger than the expected one. Expected 1 received ${updateResult.affected}`,
+          );
+        }
+
+        // Create a student notification when COE is confirmed.
+        await this.createNotificationForDisbursementUpdate(
+          disbursementScheduleId,
+          auditUserId,
+          transactionalEntityManager,
+        );
+
+        return updateResult;
+      },
+    );
+  }
+  /**
+   * Approve confirmation of enrollment(COE).
+   * An application can have up to two COEs based on the disbursement schedule,
+   * hence the COE approval happens twice for application with more than once disbursement.
+   * Irrespective of number of COEs to be approved, application status is set to complete
+   * on first COE approval.
+   * @param disbursementScheduleId disbursement schedule id of COE.
+   * @param auditUserId user who confirms enrollment.
+   * @param tuitionRemittance tuition remittance amount.
+   * @param options Confirm COE options.
+   * - `locationId` location id of the application.
+   * - `allowOutsideCOEApprovalPeriod` allow COEs which are outside the valid COE confirmation period to be confirmed..
+   * - `enrolmentConfirmationDate` date of enrolment confirmation.
+   * - `applicationNumber` application number of the enrolment.
+   */
+  async confirmEnrollment(
+    disbursementScheduleId: number,
+    auditUserId: number,
+    tuitionRemittance: number,
+    options?: {
+      locationId?: number;
+      allowOutsideCOEApprovalPeriod?: boolean;
+      enrolmentConfirmationDate?: Date;
+      applicationNumber?: string;
+    },
+  ): Promise<void> {
+    // Get the disbursement and application summary for COE.
+    const disbursementSchedule =
+      await this.getDisbursementAndApplicationSummary(
+        disbursementScheduleId,
+        options?.locationId,
+      );
+
+    if (!disbursementSchedule) {
+      throw new CustomNamedError("Enrolment not found.", ENROLMENT_NOT_FOUND);
+    }
+
+    const application = disbursementSchedule.studentAssessment.application;
+
+    if (
+      options?.applicationNumber &&
```

Comment:
do we need the first check, `options?.applicationNumber &&`?.

Reply:
Yes because application number validation is only for ECE Integration.